File: game/coach/panel.py
"""
TypingPanel — the results-screen "Typing" view (Tab).

One picture, three sentences, one thing to try: the keyboard heatmap tinted
by accuracy, Noki with a speech bubble, and the instruction line.  Reads the
folded stats and the tip library; writes the run into the local history so
next time the praise and the cooldowns are honest.
"""
from __future__ import annotations

import json
import os

import pygame

from .. import keyboard as KB
from ..settings import config_dir, load_settings
from ..sprites import get_font, render_text, load_noki_frames, multiply_alpha
from .stats import fold, merge
from .tips import build_tips, pick

_ROWS = ["qwertyuiop", "asdfghjkl", "zxcvbnm"]
_ROW_INDENT = [0.0, 0.5, 1.0]


def history_path() -> str:
    return os.path.join(config_dir(), "typing_history.json")


def load_history() -> dict:
    try:
        with open(history_path(), "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {}


def save_history(h: dict) -> None:
    try:
        tmp = history_path() + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(h, f)
        os.replace(tmp, history_path())
    except Exception:
        pass


def record_run(hits: list, stats_run: dict | None = None) -> tuple[dict, dict, list[str], dict | None]:
    """Fold a run, merge into history, decide the sentences. Returns (run_stats, history, sentences, instruction)."""
    run = stats_run or fold(hits)
    hist = load_history()
    settings = load_settings()
    tips = build_tips(run, settings, hist)
    cooldown = {k: v for k, v in hist.get("cooldown", {}).items() if v > 0}
    sentences, instruction = pick(tips, cooldown)
    # advance cooldowns and set new ones for the sentences shown
    new_cd = {k: v - 1 for k, v in cooldown.items() if v - 1 > 0}
    for t in tips:
        if t["text"] in sentences and t["priority"] <= 4:
            new_cd[t["id"]] = 3
    hist = merge(hist, run)
    hist["cooldown"] = new_cd
    save_history(hist)
    return run, hist, sentences, instruction


class TypingPanel:
    def __init__(self, screen: pygame.Surface, hits: list, stats: dict, level_name: str, difficulty: str) -> None:
        self.screen = screen
        self.level_name = level_name
        self.difficulty = difficulty
        self.run, self.history, self.sentences, self.instruction = record_run(hits)
        self.t = 0.0
        sw, sh = screen.get_size()
        self.sw, self.sh = sw, sh
        self.s = min(sw / 1920, sh / 1080)
        self.tile = int(64 * self.s)
        self.pitch = int(76 * self.s)
        self.kb_x = int(160 * self.s)
        self.kb_y = int(400 * self.s)
        self.noki = load_noki_frames("noki_bop", int(300 * self.s))
        self._dim = pygame.Surface((sw, sh), pygame.SRCALPHA)
        self._dim.fill((4, 4, 10, 235))

    def _tile_color(self, k: str) -> tuple:
        d = self.run["keys"].get(k)
        if d is None or d["presses"] < 6:
            return (58, 58, 72)
        acc = d["accuracy"]
        lane = KB.lane_color(KB.lane_of(k))
        if acc >= 0.95:
            return (240, 240, 250)
        if acc >= 0.85:
            return tuple(int(c * 0.45 + 40) for c in lane)
        if acc >= 0.75:
            return tuple(int(c * 0.85) for c in lane)
        return KB.MISS_RED

    def draw(self, dt: float) -> None:
        self.t += dt
        scr = self.screen
        s = self.s
        scr.blit(self._dim, (0, 0))
        title = render_text("display", int(30 * s), f"TYPING · {self.level_name}", (230, 230, 240))
        scr.blit(title, (self.kb_x, int(300 * s)))
        # keyboard: tiles pop in row by row
        pop_i = 0
        for r, row in enumerate(_ROWS):
            for c, k in enumerate(row):
                delay = pop_i * 0.032
                pop_i += 1
                if self.t < delay:
                    continue
                k_t = min(1.0, (self.t - delay) / 0.12)
                sc = 0.6 + 0.4 * (1 - (1 - k_t) ** 3)
                x = self.kb_x + int((c + _ROW_INDENT[r]) * self.pitch)
                y = self.kb_y + r * self.pitch
                size = int(self.tile * sc)
                rect = pygame.Rect(0, 0, size, size)
                rect.center = (x + self.tile // 2, y + self.tile // 2)
                col = self._tile_color(k)
                pygame.draw.rect(scr, col, rect, border_radius=int(10 * s))
                d = self.run["keys"].get(k)
                ink = (20, 20, 30) if sum(col) > 400 else (240, 240, 250)
                letter = render_text("display", int(26 * s), k.upper(), ink)
                scr.blit(letter, letter.get_rect(center=(rect.centerx, rect.centery - int(8 * s))))
                if d is not None and d["presses"] >= 6:
                    pct = render_text("body", int(14 * s), f"{int(d['accuracy'] * 100)}", ink)
                    scr.blit(pct, pct.get_rect(center=(rect.centerx, rect.centery + int(14 * s))))
                    if abs(d["relative_late"]) >= 20 and d["n_timed"] >= 3:
                        arrow = "▲" if d["relative_late"] < 0 else "▼"
                        a = render_text("body", int(12 * s), arrow, ink)
                        scr.blit(a, (rect.right - a.get_width() - int(4 * s), rect.top + int(3 * s)))
                    slips = min(5, d["slips_expected"])
                    for i in range(slips):
                        pygame.draw.circle(scr, KB.MISS_RED, (rect.left + int(8 * s) + i * int(9 * s), rect.bottom - int(7 * s)), int(3 * s))
        # Noki + bubble
        bx = int(1180 * s)
        if self.noki:
            n = len(self.noki)
            frame = self.noki[int((self.t * 2.2) % 1.0 * n) % n]
            scr.blit(frame, (int(1520 * s), int(360 * s)))
        by = int(420 * s)
        bubble_w = int(330 * s)
        lines: list[pygame.Surface] = []
        font = get_font("display_regular", int(24 * s))
        for i, sent in enumerate(self.sentences):
            if self.t < 0.9 + i * 0.5:
                break
            for chunk in self._wrap(sent, font, bubble_w - int(32 * s)):
                lines.append(font.render(chunk, True, (20, 20, 30)))
        if lines:
            h = sum(l.get_height() for l in lines) + int(28 * s)
            rect = pygame.Rect(bx, by, bubble_w, h)
            pygame.draw.rect(scr, (245, 245, 250), rect, border_radius=int(18 * s))
            pygame.draw.polygon(scr, (245, 245, 250), [(rect.right - int(20 * s), rect.centery - int(10 * s)),
                                                        (rect.right + int(18 * s), rect.centery),
                                                        (rect.right - int(20 * s), rect.centery + int(10 * s))])
            yy = rect.top + int(14 * s)
            for l in lines:
                scr.blit(l, (rect.left + int(16 * s), yy))
                yy += l.get_height()
        if self.instruction and self.t > 0.9 + len(self.sentences) * 0.5:
            lab = render_text("stamp", int(18 * s), "TRY THIS", KB.GOLD)
            scr.blit(lab, (bx, int(720 * s)))
            txt = render_text("body", int(24 * s), self.instruction["try"], (230, 230, 240))
            scr.blit(txt, (bx, int(750 * s)))
        # strip
        run = self.run
        weak_f = None
        worst = 1.0
        for f, d in run["fingers"].items():
            if d["presses"] >= 8 and d["accuracy"] < worst:
                worst, weak_f = d["accuracy"], f
        parts = []
        if weak_f is not None:
            parts.append(f"Weakest finger: {KB.FINGER_NAMES[weak_f]}")
        if run["wpm"] > 0:
            parts.append(f"Pace: {int(round(run['wpm']))} words a minute (set by the song)")
        parts.append(f"Slips {run['slips']}")
        strip = render_text("body", int(18 * s), "   ·   ".join(parts), (150, 155, 175))
        scr.blit(strip, (self.kb_x, self.kb_y + 3 * self.pitch + int(30 * s)))
        hint = render_text("body", int(16 * s), "TAB or ESC · back", (110, 115, 140))
        scr.blit(hint, hint.get_rect(bottomright=(self.sw - int(40 * s), self.sh - int(30 * s))))

    @staticmethod
    def _wrap(text: str, font: pygame.font.Font, width: int) -> list[str]:
        words = text.split()
        lines: list[str] = []
        cur = ""
        for w in words:
            trial = (cur + " " + w).strip()
            if font.size(trial)[0] <= width or not cur:
                cur = trial
            else:
                lines.append(cur)
                cur = w
        if cur:
            lines.append(cur)
        return lines

File: web/src/coach/panel.ts
/**
 * The results-screen "Typing" view (Tab).  A port of game/coach/panel.py.
 *
 * One picture, three sentences, one thing to try: the keyboard heatmap tinted by
 * accuracy, Noki with a speech bubble, and the instruction line.  Reads the
 * folded stats and the tip library; writes the run into local history so next
 * time the praise and the cooldowns are honest.
 *
 * The pygame panel draws its own keyboard tile by tile with a pop-in animation.
 * Here the keyboard is a CSS grid and the pop-in is one keyframe, so the whole
 * view is markup — which also means the numbers on it are selectable and the
 * sentences are read by a screen reader.
 */
import * as KB from '../core/keyboard'
import type { HitRecord } from '../core/models'
import { fold, merge } from './stats'
import type { History, RunStats } from './stats'
import { buildTips, pick } from './tips'
import type { Tip } from './tips'

const ROWS = ['qwertyuiop', 'asdfghjkl', 'zxcvbnm']
const ROW_INDENT = [0, 0.5, 1]
const HISTORY_KEY = 'noki.typing_history'

export function loadHistory(): History {
  try {
    return JSON.parse(localStorage.getItem(HISTORY_KEY) ?? '{}') as History
  } catch {
    return {}
  }
}

export function saveHistory(h: History): void {
  try {
    localStorage.setItem(HISTORY_KEY, JSON.stringify(h))
  } catch {
    /* private window, or storage disabled: the coach just forgets */
  }
}

export interface RecordedRun {
  run: RunStats
  history: History
  sentences: string[]
  instruction: Tip | null
}

/** Fold a run, merge it into history, decide the sentences. */
export function recordRun(
  hits: HitRecord[],
  settings: Record<string, unknown> = {},
  runStats?: RunStats,
): RecordedRun {
  const run = runStats ?? fold(hits)
  let hist = loadHistory()
  const tips = buildTips(run, settings, hist)
  const cooldown: Record<string, number> = {}
  for (const [k, v] of Object.entries(hist.cooldown ?? {})) if (v > 0) cooldown[k] = v
  const [sentences, instruction] = pick(tips, cooldown)
  // advance cooldowns and set new ones for the sentences shown
  const newCd: Record<string, number> = {}
  for (const [k, v] of Object.entries(cooldown)) if (v - 1 > 0) newCd[k] = v - 1
  for (const t of tips) if (sentences.includes(t.text) && t.priority <= 4) newCd[t.id] = 3
  hist = merge(hist, run)
  hist.cooldown = newCd
  saveHistory(hist)
  return { run, history: hist, sentences, instruction }
}

/** Tile colour by accuracy, matching TypingPanel._tile_color. */
function tileColor(run: RunStats, k: string): string {
  const d = run.keys[k]
  if (d === undefined || d.presses < 6) return 'rgb(58,58,72)'
  const acc = d.accuracy
  const lane = KB.laneColor(KB.laneOf(k))
  const r = (lane >> 16) & 255
  const g = (lane >> 8) & 255
  const b = lane & 255
  if (acc >= 0.95) return 'rgb(240,240,250)'
  if (acc >= 0.85) {
    return `rgb(${Math.trunc(r * 0.45 + 40)},${Math.trunc(g * 0.45 + 40)},${Math.trunc(b * 0.45 + 40)})`
  }
  if (acc >= 0.75) return `rgb(${Math.trunc(r * 0.85)},${Math.trunc(g * 0.85)},${Math.trunc(b * 0.85)})`
  const m = KB.MISS_RED
  return `rgb(${(m >> 16) & 255},${(m >> 8) & 255},${m & 255})`
}

function isLight(css: string): boolean {
  const m = css.match(/\d+/g)
  if (!m) return false
  return Number(m[0]) + Number(m[1]) + Number(m[2]) > 400
}

/** Build the Typing panel as a detached element. */
export function buildCoachPanel(
  hits: HitRecord[],
  levelName: string,
  settings: Record<string, unknown>,
  onClose: () => void,
): HTMLElement {
  const { run, sentences, instruction } = recordRun(hits, settings)

  const el = document.createElement('div')
  el.className = 'overlay coach'
  el.setAttribute('role', 'dialog')
  el.setAttribute('aria-label', `Typing report for ${levelName}`)

  // one running index across all three rows, so the pop-in sweeps the keyboard
  // rather than restarting each row (game/coach/panel.py's pop_i)
  let popI = 0
  const kb = ROWS.map((row, r) => {
    const keys = [...row].map((k) => {
      const i = popI++
      const d = run.keys[k]
      const col = tileColor(run, k)
      const ink = isLight(col) ? '#14141e' : '#f0f0fa'
      const shown = d !== undefined && d.presses >= 6
      const pct = shown ? `<span class="pct">${Math.trunc(d.accuracy * 100)}</span>` : ''
      // an arrow only where the key is clearly off the run's own median
      const arrow = shown && Math.abs(d.relative_late) >= 20 && d.n_timed >= 3
        ? `<span class="arrow">${d.relative_late < 0 ? '▲' : '▼'}</span>` : ''
      const slips = shown
        ? Array.from({ length: Math.min(5, d.slips_expected) }, () => '<i></i>').join('')
        : ''
      const title = shown
        ? `${k.toUpperCase()}: ${Math.trunc(d.accuracy * 100)}% of ${d.presses}, ${d.slips_expected} slips`
        : `${k.toUpperCase()}: too few presses to judge`
      return `<div class="key" style="--bg:${col};--ink:${ink};--i:${i}" title="${title}">
          <span class="ltr">${k.toUpperCase()}</span>${pct}${arrow}
          <span class="slips">${slips}</span>
        </div>`
    }).join('')
    return `<div class="krow" style="--indent:${ROW_INDENT[r]}">${keys}</div>`
  }).join('')

  // the strip: weakest finger, pace, slips
  let weakF: number | null = null
  let worst = 1
  for (const [fs, d] of Object.entries(run.fingers)) {
    if (d.presses >= 8 && d.accuracy < worst) { worst = d.accuracy; weakF = Number(fs) }
  }
  const parts: string[] = []
  if (weakF !== null) parts.push(`Weakest finger: ${KB.FINGER_NAMES[weakF]}`)
  if (run.wpm > 0) parts.push(`Pace: ${Math.round(run.wpm)} words a minute (set by the song)`)
  parts.push(`Slips ${run.slips}`)

  el.innerHTML = `
    <div class="coach-card">
      <div class="label">Typing · ${escapeHtml(levelName)}</div>
      <div class="coach-body">
        <div class="keyboard" aria-label="Accuracy by key">${kb}</div>
        <div class="bubble-col">
          <div class="bubble-row">
            <div class="bubble">${sentences.map((s) => `<p>${escapeHtml(s)}</p>`).join('')}</div>
            <div class="coach-noki" role="presentation"></div>
          </div>
          ${instruction && instruction.try
            ? `<div class="tryit"><span class="label">Try this</span><p>${escapeHtml(instruction.try)}</p></div>`
            : ''}
        </div>
      </div>
      <div class="strip">${escapeHtml(parts.join('   ·   '))}</div>
      <button class="play" id="coach-back">Back</button>
    </div>
  `
  ;(el.querySelector('#coach-back') as HTMLButtonElement).onclick = onClose
  return el
}

function escapeHtml(s: string): string {
  return s.replace(/[&<>"']/g, (c) =>
    ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]!))
}
